### PR TITLE
Some JSInterpreter code cleanup

### PR DESCRIPTION
**Pull request by pcardune**
- Moved JSInterpreter to lib/tools/jsinterpreter/JSInterpreter.js where it belongs
- Switch everything to es6 class syntax
- Removed use of unnecessary bind calls in favor of arrow functions
- Remove direct access to interpreter stack, putting everything behind accessor functions instead
- Replace var with const and let throughout

### CODE REVIEW

File: apps/src/codegen.js [hunk 1]
```diff
@@ -92,8 +93,11 @@ export function evalWithEvents(apis, events, evalCode = '') {
   // to call, and any arguments.
   const eventLoop = ';while(true){var event=wait();setReturnValue(this[event.name].apply(null,event.args));}';
 
-  interpreter = new Interpreter(
+  // TODO (pcardune): remove circular dependency
```
Comment (islemaster) [inline apps/src/codegen.js:96]: When will this TODO be addressed?

File: apps/src/turtle/turtle.js [hunk 2]
```diff
@@ -276,8 +276,8 @@ Artist.prototype.init = function (config) {
  */
 Artist.prototype.prepareForRemix = function () {
   if (REMIX_PROPS.every(group => Object.keys(group.defaultValues).every(prop =>
-        this.level[prop] === undefined ||
-            this.level[prop] === group.defaultValues[prop]))) {
+    this.level[prop] === undefined ||
+                                                                              this.level[prop] === group.defaultValues[prop]))) {
```
Comment (islemaster) [inline apps/src/turtle/turtle.js]: This whitespace change looks like a mistake.

File: apps/src/lib/tools/jsinterpreter/JSInterpreter.js [hunk 3]
```diff
@@ -0,0 +1,1096 @@
...
+    }
+  }
+
+  /**
+   * Determines if the program is done executing.
+   *
+   * @return {boolean} true if program is complete (or an error has occurred).
+   */
+  isProgramDone() {
+    const topStackFrame = this.interpreter.peekStackFrame();
```
Comment (islemaster) [inline apps/src/lib/tools/jsinterpreter/JSInterpreter.js]: Possible null reference: Should probably pull the `this.interpreter` null check in the expression below up into this assignment.

File: apps/src/lib/tools/jsinterpreter/JSInterpreter.js [hunk 4]
```diff
@@ -0,0 +1,1096 @@
...
+   * optionally, callback arguments (stored in "arguments")
+   */
+  nativeGetCallback = () => {
+    this.startedHandlingEvents = true;
+    const retVal = this.eventQueue.shift();
+    if (typeof retVal === "undefined") {
+      this.yield();
+    }
+    return retVal;
+  }
```
Comment (islemaster) [inline apps/src/lib/tools/jsinterpreter/JSInterpreter.js]: Missing semicolon.

File: apps/src/lib/tools/jsinterpreter/JSInterpreter.js [hunk 5]
```diff
@@ -0,0 +1,1096 @@
...
+    if (!this.runUntilCallbackReturn &&
+                                     typeof this.lastCallbackRetVal !== 'undefined') {
+      this.onExecutionWarning.notifyObservers("Function passed to onEvent() " +
+                                              "has taken too long - the return value was ignored.");
+      if (!this.shouldRunAtMaxSpeed()) {
+        this.onExecutionWarning.notifyObservers("  (try moving the speed " +
+                                                "slider to its maximum value)");
+      }
+    }
+  }
```
Comment (islemaster) [inline apps/src/lib/tools/jsinterpreter/JSInterpreter.js]: Missing semicolon.

File: apps/src/lib/tools/jsinterpreter/JSInterpreter.js [hunk 6]
```diff
@@ -0,0 +1,1096 @@
...
+        this.maxValidCallExpressionDepth = stackDepth;
+
+        if (inUserCode && this.interpreter.peekStackFrame().node.type === "CallExpression") {
+          // Store that we've seen a call expression at this depth in callExpressionSeenAtDepth:
+          this.callExpressionSeenAtDepth[stackDepth] = true;
+        }
+
+        if (this.paused) {
+          // Store the first call expression stack depth seen while in this step operation:
+          if (inUserCode && this.interpreter.peekStackFrame().node.type === "CallExpression") {
```
Comment (islemaster) [inline apps/src/lib/tools/jsinterpreter/JSInterpreter.js]: I still think extracting `inUserCallExpression` (here and eight lines above) makes sense.

File: apps/src/codegen.js (hunk 1 above)
Comment (pcardune) [reply apps/src/codegen.js:96]: I need to move `codegen` to `lib/tools/jsinterpreter`. At that point, I will probably move a number of these functions onto CustomMarshalingInterpreter and kill off the circular deps.

File: apps/src/turtle/turtle.js (hunk 2 above)
Comment (pcardune) [reply apps/src/turtle/turtle.js]: yep. I blame my editor which does auto-indenting.

File: apps/src/lib/tools/jsinterpreter/JSInterpreter.js (hunk 5 above)
Comment (pcardune) [reply apps/src/lib/tools/jsinterpreter/JSInterpreter.js]: interesting. eslint doesn't seem to catch this.

**Comment by pcardune** [comment]
Ok, pushed a new version without the unintentional indentation changes.

File: apps/src/lib/tools/jsinterpreter/JSInterpreter.js [hunk 7]
```diff
@@ -0,0 +1,1100 @@
...
+   *
+   * @return {boolean} true if program is complete (or an error has occurred).
+   */
+  isProgramDone() {
+    if (!this.interpreter) {
+      throw new Error("Program has not even been run yet.");
+    }
+    const topStackFrame = this.interpreter.peekStackFrame();
+    return this.executionError ||
+           !this.interpreter ||
```
Comment (islemaster) [inline apps/src/lib/tools/jsinterpreter/JSInterpreter.js]: Nit: This guard is now redundant.

**Pull request merged**